remove GtkStatusIcon code

LaserEyess opened the pull request:
GtkStatusIcon has been deprecated since 3.14, it was an implementation
of the legacy X11 systray functionality that is no longer relevant for
modern desktop environments. AppIndicators are the replacement on most
DEs, with GNOME being a notable exception. Though there is a gnome-shell
extension to add support.

Since transmission-remote-gtk already has AppIndicator support, only a
minor amount of refactoring was needed to drop support all together.


A note about my `#if`ing. I tried to keep them confined to single functions, either the entire function, or by making the function a noop if `HAVE_LIBAPPINIDATOR` is false. I think this is better than littering individual functions with multiple `#if`s. Especially since this particular refactor would requie putting new statements in `src/trg-preferences-dialog.c`.

Tested my changes with and without libappindicator.

LaserEyess commented:
Hm, just realized I will need at least one `#if` in `src/trg-preferences-dialog.c` since otherwise there is a button for the system tray that doesn't do anything.

TingPing commented on src/trg-main-window.c:
```diff
@@ -2483,42 +2402,18 @@ static void trg_main_window_set_hidden_to_tray(TrgMainWindow * win,
-#else
-    if (1) {
-#endif
-        if (priv->statusIcon)
-            g_object_unref(G_OBJECT(priv->statusIcon));
-
-        priv->statusIcon = NULL;
     }
```
While in the area:
```suggestion
    g_clear_pointer(&priv->appIndicator, g_object_unref);
```

TingPing commented:
> Still I wonder if the is_unity() function is relevant anymore...

Nope, remove it.

TingPing commented on src/trg-preferences-dialog.c:
```diff
@@ -756,7 +756,12 @@ static GtkWidget *trg_prefs_viewPage(TrgPreferencesDialog * dlg)
 #endif
 
-	hig_workarea_add_section_title(t, &row, _("System Tray"));
+#if HAVE_LIBAPPINDICATOR
```
I generally prefer the approach of keeping the section, but greying it out as insensitive with a tooltip that says the project was built without Appindicator support. It might be slightly technical but it clears up some confusion about which features actually exist. Realistically only users who build it themselves will even encounter this situation and can handle it.

LaserEyess replied:
Good idea

LaserEyess commented:
@Tingping PTAL

 - Adjusterd `connchange_whatever_tray()` to be a true noop with no libappindicator
 - fixed indentation in `trg_prefs_viewPage()`
 - Added tooltip/insensitive toggle to system tray when there's no libappindicator
 - new commit removing `is_unity()`

One thing I noticed: the actual tray icon doesn't go away on my tray (waybar). It doesn't do anything when system tray is unchecked, but the icon remains. This behavior seems to be the same in the master branch, so I don't think I introduced a bug. Is there anything special you have to do to un-init it besides clearing the object?

TingPing commented:
> I tried looking at the libappindicator source code but there's no app_indicator_tell_the_tray_its_gone() function or anything. Not a public one at least.

Along the lines of `app_indicator_set_status (ind, APP_INDICATOR_STATUS_HIDDEN);` (I say without finding any real API docs online....)

TingPing commented:
I'm confused how freeing it wasn't enough though.

TingPing commented:
Its very broken here on GNOME too. Also the Quit menu entry in it at the very least doesn't work.

TingPing commented:
Not your bug but a random note:

Creating a new menu and calling `app_indicator_set_menu(priv->appIndicator, menu);` each time is a bad design. We should update the `GtkMenu` in-place.

TingPing commented on src/trg-main-window.c:
```diff
@@ -2570,27 +2452,11 @@ void trg_main_window_add_status_icon(TrgMainWindow * win)
         app_indicator_set_status(priv->appIndicator,
```
I recommend a guard here to ensure `priv->appIndicator` is never set twice.

(It wasn't in my testing so that isn't why it won't disappear)

LaserEyess commented on src/trg-main-window.c:
```diff
@@ -2483,42 +2402,18 @@ static void trg_main_window_set_hidden_to_tray(TrgMainWindow * win,
-            g_object_unref(G_OBJECT(priv->statusIcon));
+    if (priv->appIndicator)
+        app_indicator_set_status(priv->appIndicator,
+                                 APP_INDICATOR_STATUS_PASSIVE);
```
Explicitly setting this before removing the object didn't work either. I'm out of ideas....

LaserEyess replied:
Could it be the remaining reference to the menu?

LaserEyess replied:
Nope, that didn't work either.

LaserEyess commented:
Ok, so I addressed your comment, but as for the bug with the icon not going away, that might just be a libappindicator bug or something. Maybe icons aren't supposed to go away? I can't say. I would prefer to fix both this bug, as well as the wastefulness of creating a brand new  `GtkMenu` every change, in another PR. They both exist in the current master and this PR is more about removing the deprecated `GtkStatusIcon`.

I guess it's good that we know this is broken though.
